File: java-app/src/main/java/com/mercadolibre/java_app/review/presenter/ReviewPresenter.java
package com.mercadolibre.java_app.review.presenter;

import com.mercadolibre.java_app.commons.Prices;
import com.mercadolibre.java_app.commons.Product;
import com.mercadolibre.java_app.commons.Purchase;
import com.mercadolibre.java_app.models.shipping.ShippingOption;
import com.mercadolibre.java_app.review.ReviewContract;
import com.mercadolibre.java_app.services.shipping.ShippingOptionsInteractor;

public class ReviewPresenter {
    private final ReviewContract.ReviewView reviewView;
    private final Product product;
    private final ShippingOption shippingOption;

    public ReviewPresenter(ReviewContract.ReviewView reviewView, ShippingOptionsInteractor shippingOptionsInteractor) {
        this.reviewView = reviewView;
        this.product = new Product("idProduct", 100.0);
        this.shippingOption = shippingOptionsInteractor.retrieveBestShippingOption();
    }

    public void calculatePrices() {
        showProductPrices();
        showShippingPrices();
        showTotal();
    }

    private void showProductPrices() {
        reviewView.showProductPrice(Prices.priceFormat(product.getPrice()));
    }

    private void showShippingPrices() {
        shippingOption.showShippingPrices(reviewView);
    }

    private void showTotal() {
        reviewView.showTotalPrice(shippingOption.calculatePrice(product.getPrice()));
    }

    public void validatePurchase() {
        if (product != null && shippingOption != null) {
            Purchase purchase = new Purchase(product, shippingOption);
            reviewView.goToCongrats(purchase);
        }
    }
}

File: kt-app/src/main/java/com/mercadolibre/melikotlinmeetup/review/presenter/ReviewPresenter.kt
package com.mercadolibre.melikotlinmeetup.review.presenter

import com.mercadolibre.melikotlinmeetup.commons.Product
import com.mercadolibre.melikotlinmeetup.commons.Purchase
import com.mercadolibre.melikotlinmeetup.commons.ShippingOption
import com.mercadolibre.melikotlinmeetup.commons.toPriceFormat
import com.mercadolibre.melikotlinmeetup.review.ReviewView
import com.mercadolibre.melikotlinmeetup.services.shipping.ShippingOptionsInteractor

class ReviewPresenter(private var reviewView: ReviewView, private val shippingOptionsInteractor: ShippingOptionsInteractor) {
    private val shippingOption: ShippingOption?
    private val product: Product?

    init {
        this.shippingOption = shippingOptionsInteractor.retrieveBestShippingOption()
        this.product = Product("idProduct", 100.0)
    }

    fun calculatePrices() {
        showProductPrices()
        showShippingPrices()
        showTotal()
    }

    private fun showProductPrices() {
        reviewView.showProductPrice(product?.price!!.toPriceFormat())
    }

    private fun showShippingPrices() {
        shippingOption?.showShippingPrices(reviewView)
    }

    private fun showTotal() {
        reviewView.showTotalPrice(shippingOption!!.calculatePrice(product?.price!!))
    }

    fun validatePurchase() {
        if (product != null && shippingOption != null) {
            val purchase = Purchase(product, shippingOption)
            reviewView.goToCongrats(purchase)
        }
    }
}